### PR TITLE
Do relocations after app's flash memory is moved

### PR DESCRIPTION
**Pull request by dfordivam**

### CODE REVIEW

**Comment by siy** [comment]
Tried this fix for my app, but it fails with following error:

```
          scripts/link_wrap.sh: line 23: OBJCOPY: unbound variable
```
Ready to provide full link output (quite long) if necessary.

**Comment by Ericson2314** [comment]
@siy is your app public? Needing `OBJCOPY` to be defined was true both before and after this PR.

File: src/c/src.c [hunk 1]
```diff
@@ -162,6 +168,76 @@ void link_pass(
   /* PRINTLNC("Ending link pass"); */
 }
 
+void link_pass_ram(
+  size_t sec_len,
+  struct SectionSrc *sec_src,
+  struct SectionDst *sec_dst)
+{
+    link_pass(sec_len, sec_src, sec_dst, 0, NULL, NULL, true);
```
Comment (Ericson2314) [inline src/c/src.c]:
```suggestion
#if defined(ST31)
  SYMBOL_ABSOLUTE_VALUE(nvram_ptr, _nvram);
  SYMBOL_ABSOLUTE_VALUE(envram_ptr, _envram);
#elif defined(ST33) || defined(ST33K1M5)
  __asm volatile("ldr %0, =_nvram":"=r"(nvram_ptr));
  __asm volatile("ldr %0, =_envram":"=r"(envram_ptr));
#else
#error "invalid architecture"
#endif

  // Value of _nvram in this run
  void* nvram_current = pic(nvram_ptr);

  link_pass(sec_len, sec_src, sec_dst, pic(nvram_ptr) - nvram_ptr, nvram_ptr, envram_ptr true);
```

File: src/c/src.c [hunk 2]
```diff
@@ -138,10 +140,14 @@ void link_pass(
       // Assuming no relocations go behind the end address.
       if (word_offset < sizeof(buf) / sizeof(*buf)) {
         PRINTLNC("Possible reloc");
-        link_addr_t old = buf[word_offset];
-        install_addr_t new = pic(old);
+        void* old = (void*) buf[word_offset];
+        void* new = pic(old);
+        if (old == new && nvram_move_amt != 0 && old >= nvram_prev && old < envram_prev) {
```
Comment (Ericson2314) [inline src/c/src.c]:
```suggestion
        if (nvram_move_amt != 0 && old >= nvram_prev && old < envram_prev) {
```

File: src/c/src.c [hunk 3]
```diff
@@ -162,6 +168,76 @@ void link_pass(
...
+      // No change in _nvram means that we need not do link_pass again
+      return;
+  }
+
+  // Value (in bytes) of change in _nvram
+  // If the app was moved after the previous run
+  int nvram_move_amt = 0;
+  if (nvram_prev != NULL) {
+      nvram_move_amt = nvram_current - nvram_prev;
+  }
```
Comment (Ericson2314) [inline src/c/src.c]:
```suggestion
  int nvram_move_amt = nvram_current - nvram_prev;
```

File: link.ld [hunk 4]
```diff
@@ -85,6 +85,13 @@ SECTIONS
     _nvm_data = .;
     _nvm_data_src = .;
     *(.nvm_data*)
+
+    /* Store _nvram value during link_pass and use this to detect movement of
+    _nvram as compared to the previous app execution, and redo the relocations
+    if necessary */
+    _nvram_prev_run = .;
+    . += 4;
```
Comment (Ericson2314) [inline link.ld]: setting value here as we discussed.

File: src/c/src.c (hunk 1 above)
Comment (Ericson2314) [reply src/c/src.c]: And then good to deduplicate defining `nvram_ptr` and `envram_ptr` with the other function somehow.

**Comment by siy** [comment]
> @siy is your app public? Needing `OBJCOPY` to be defined was true both before and after this PR.

No, app is not public yet. OK, I'll try to fix it.

**Comment by siy** [comment]
After adding OBJCOPY env:
```
          + arm-none-eabi-objcopy --dump-section .rel.rodata=/home/siy/CLionProjects/babylon-ledger-app/target/nanosplus/debug/deps/babylon_ledger_app-80f66e0e75e9e7f0-rodata-reloc /home/siy/CLionProjects/babylon-ledger-app/target/nanosplus/debug/deps/babylon_ledger_app-80f66e0e75e9e7f0 /dev/null
          arm-none-eabi-objcopy: /home/siy/CLionProjects/babylon-ledger-app/target/nanosplus/debug/deps/babylon_ledger_app-80f66e0e75e9e7f0: can't dump section '.rel.rodata' - it does not exist: file format not recognized
          arm-none-eabi-objcopy: /dev/null: error: PHDR segment not covered by LOAD segment
          arm-none-eabi-objcopy: /dev/null[.text]: file format not recognized
          RUST_LLD DONE
          OUT IS /home/siy/CLionProjects/babylon-ledger-app/target/nanosplus/debug/deps/babylon_ledger_app-80f66e0e75e9e7f0
```

**Comment by dfordivam** [comment]
@siy this is unfortunately an issue with (some versions of) the GNU arm-none-eabi-objcopy, 
This should work with llvm's objcopy

**Comment by Ericson2314** [comment]
@siy If you are using Nix with Alamgu all the right tools will be gotten for you and it should "just work".

**Comment by siy** [comment]
> @siy If you are using Nix with Alamgu all the right tools will be gotten for you and it should "just work".

I'm not using Nix nor Alamgu. I'm using Ledger Rust SDK.

**Pull request merged**